pkg/traces/promsdprocessor/consumer/consumer.go

```diff
@@ -0,0 +1,254 @@
+			NextConsumer:    nextConsumer,
+		},
+		logger: logger,
+	}, nil
+}
+
+// UpdateOptions is used in flow mode, where all options need to be updated.
+func (c *Consumer) UpdateOptions(opts Options) {
+	c.optsMut.Lock()
+	defer c.optsMut.Unlock()
+
+	c.opts = opts
+}
+
+// UpdateOptionsHostLabels is used in static mode, where only the host labels need to be updated.
+func (c *Consumer) UpdateOptionsHostLabels(hostLabels map[string]discovery.Target) {
+	c.optsMut.Lock()
+	defer c.optsMut.Unlock()
+
+	c.opts.HostLabels = hostLabels
+}
+
+func (c *Consumer) ConsumeTraces(ctx context.Context, td ptrace.Traces) error {
+	c.optsMut.RLock()
+	defer c.optsMut.RUnlock()
+
+	rss := td.ResourceSpans()
+	for i := 0; i < rss.Len(); i++ {
+		rs := rss.At(i)
+
+		c.processAttributes(ctx, rs.Resource().Attributes())
+	}
+
+	return c.opts.NextConsumer.ConsumeTraces(ctx, td)
+}
+
+func (c *Consumer) Capabilities() otelconsumer.Capabilities {
+	return otelconsumer.Capabilities{MutatesData: true}
+}
+
+func (c *Consumer) processAttributes(ctx context.Context, attrs pcommon.Map) {
+	ip := c.getPodIP(ctx, attrs)
+	// have to have an ip for labels lookup
+	if ip == "" {
+		level.Debug(c.logger).Log("msg", "unable to find ip in span attributes, skipping attribute addition")
+		return
+	}
+
+	labels, ok := c.opts.HostLabels[ip]
+	if !ok {
+		level.Debug(c.logger).Log("msg", "unable to find labels", "ip", ip)
+		return
+	}
+
+	for k, v := range labels {
+		switch c.opts.OperationType {
+		case OperationTypeUpsert:
+			attrs.PutStr(k, v)
+		case OperationTypeInsert:
+			if _, ok := attrs.Get(k); !ok {
+				attrs.PutStr(k, v)
+			}
+		case OperationTypeUpdate:
+			if toVal, ok := attrs.Get(k); ok {
+				toVal.SetStr(v)
+			}
+		}
+	}
+}
+
+func (c *Consumer) getPodIP(ctx context.Context, attrs pcommon.Map) string {
+	for _, podAssociation := range c.opts.PodAssociations {
+		switch podAssociation {
+		case PodAssociationIPLabel, PodAssociationOTelIPLabel, PodAssociationk8sIPLabel:
+			ip := stringAttributeFromMap(attrs, podAssociation)
+			if ip != "" {
+				return ip
+			}
+		case PodAssociationHostnameLabel:
+			hostname := stringAttributeFromMap(attrs, semconv.AttributeHostName)
+			if net.ParseIP(hostname) != nil {
+				return hostname
+			}
+		case PodAssociationConnectionIP:
+			ip := c.getConnectionIP(ctx)
+			if ip != "" {
+				return ip
+			}
+		}
+	}
+	return ""
+}
+
+func stringAttributeFromMap(attrs pcommon.Map, key string) string {
+	if attr, ok := attrs.Get(key); ok {
+		if attr.Type() == pcommon.ValueTypeStr {
+			return attr.Str()
+		}
+	}
+	return ""
+}
+
+func (c *Consumer) getConnectionIP(ctx context.Context) string {
+	cl := client.FromContext(ctx)
+	if cl.Addr == nil {
+		return ""
+	}
+
+	host := cl.Addr.String()
+	if strings.Contains(host, ":") {
+		var err error
+		splitHost, _, err := net.SplitHostPort(host)
+		if err != nil {
+			// It's normal for this to fail for IPv6 address strings that don't actually include a port.
```

Comment:
If its normal should we return an error at all? IE only error if IP4?

Reply:
I think this is why this is a debug log instead of a warning log. It's supposed to be ok for a failure to happen, but the debug log might be of interest in case it's a genuinely malformed hostname (e.g. `::1]:80` instead of  `[::1]:80`). Btw this function is from [old code](https://github.com/grafana/agent/blob/v0.35.3/pkg/traces/promsdprocessor/prom_sd_processor.go#L122-L141).

Reply:
Btw we are not returning an error. In case of an error, we are returning the original address string.